[BugFix] Error in documentation's numpy differentiation code example (#1499)
PennyLaneAI/pennylane: doc/introduction/interfaces/numpy.rst

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

How does PennyLane know which arguments of a quantum function to differentiate, and which to ignore?

> For example, you may want to pass arguments to a QNode but *not* have

Author: I decided to polish the entire paragraph, so it makes sense together. I think here `as positional arguments` is confusing because we only state the rule about positional arguments below.

PennyLane consider them when computing gradients.

As a basic rule, **all positional arguments provided to the QNode are assumed to be differentiable

> by default**. To accomplish this, arguments are internally turned into arrays of the PennyLane NumPy module,

Author: I think this piece of information is super important!

    The ``requires_grad`` argument can be passed to any NumPy function provided by PennyLane,
    including NumPy functions that create arrays like ``np.random.random``, ``np.zeros``, etc.

> An alternative way to avoid differentiation logic to be applied to positional arguments is to

Author: ```suggestion
An alternative way to avoid having positional arguments turned into differentiable PennyLane NumPy arrays is to
```

    dev = qml.device('default.qubit', wires=5)

    @qml.qnode(dev)

>     def circuit(weights, data, wires=None):

Reviewer: ```suggestion
    def circuit(weights, data, wires):
```

This can stay as is --- the workaround is just that it needs to be _passed_ as a kwarg!

        qml.CNOT(wires=[wires[0], wires[2]])
        return qml.expval(qml.PauliZ(wires[0]))

> Since ``wires`` is a keyword argument, it will be automatically non-trainable. However, we

Author: @josh146 will this also be true after we give up the backwards compatibility hack?

Reviewer: Nope! In fact, this isn't even a PL thing, but an autograd thing - it is autograd that is ignoring keyword arguments 😆 So maybe best to reword this?

> must make ``data`` non-trainable by specifying it as a NumPy array with ``requires_grad=False``:

Author: ```suggestion
must explicitly make ``data`` non-trainable by specifying it as a NumPy array with ``requires_grad=False``:
```

>>> np.random.seed(42)  # make the results reproducable
>>> weights = np.array([0.1, 0.2, 0.3])
>>> data = np.random.random([2**3], requires_grad=False)
>>> wires = [2, 0, 1]

> >>> circuit(weights, data, wires=wires)

Reviewer: As far as I am aware, this is the only code change necessary to make it work, the change above (`wires=None`) should not be needed

0.4124409353413991

When we compute the derivative, arguments with ``requires_grad=False``, as well as keyword arguments,

> are ignored by :func:`~.grad`:

Reviewer: 👍 

I think this is all that needs to be said?

> are explicitly ignored by :func:`~.grad`:

Author: ```suggestion
are ignored by :func:`~.grad`:
```

        qml.CNOT(wires=[wires[0], wires[2]])
        return qml.expval(qml.PauliZ(wires[0]))

For ``data``, which is a PennyLane NumPy array, we can simply specify ``requires_grad=False``:

>>> np.random.seed(42)  # make the results reproducable
>>> data = np.random.random([2**3], requires_grad=False)

> But ``wires`` is a list in this example, and if we turned it into a PennyLane NumPy array we would have to

Author: ```suggestion
But ``wires`` is a list in this example, and if we turn it into a PennyLane NumPy array we would have to
```
